### PR TITLE
Maintenance schedules system assignment UI

### PR DESCRIPTION
**Pull request by cbbayburt**
Adds UI for assigning maintenance schedules to systems.

### TODO
 - [x] Add an info text to the SSM panel header
 - [x] Add "Schedule" tab and schedule assignment form to SSM interface
 - [x] Add "Assign Systems" tab to the schedule details page to pick multiple systems to assign to a specific schedule
 - [x] Implement schedule assignment UI in the system properties page
 - [x] Add a shortcut link to the SSM summary page
 - [x] Move the SSM tab under "Misc" > "Maintenance Windows"
 - [x] Implement server side pagination for the systems list table (https://github.com/SUSE/spacewalk/issues/11895)
 - [x] Elaborate the system list with assigned schedule names (https://github.com/SUSE/spacewalk/issues/11895)
 - [x] Add confirm dialogs for when canceling affected actions
 - [x] Unify the notification messages (~`toastr`~ or regular)
 - [x] Add ~a flag to cancel affected actions~ `strategies` parameter to `assignScheduleToSystems` XMLRPC endpoint
 - [x] Test user/org permissions
 - [x] Organize the code structure

## GUI diff
*Assignment from schedule details page*
mw-schedules

*Assignment from SSM*
mw-ssm

*Cancel affected actions confirmation*
mw-ssm-confirm

*Assignment from system properties page*
mw-system-properties

## Documentation
- Will be covered as part of the "Maintenance Windows" feature

## Test coverage
- Unit tests were adapted

## Links

Fixes https://github.com/SUSE/spacewalk/issues/11228

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java [hunk 1]
```diff
@@ -115,6 +118,22 @@ public static ModelAndView maintenanceSchedules(Request request, Response respon
...
 
+    /**
+     * Handler for the SSM schedule assignment page.
+     *
+     * @param request the request object
+     * @param response the response object
+     * @param user the current user
+     * @return the ModelAndView object to render the page
+     */
+    public static ModelAndView ssmSchedules(Request request, Response response, User user) {
```
Comment (parlt91) [inline java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java:129]: I wonder whether we should move this one to the `SsmController` instead. At least that's where I would expect this to be found.

File: web/html/src/manager/maintenance/ssm/schedule-picker.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+type WithMaintenanceSchedulesProps = {
+  systems: string[],
+  onMessage: (messages: MessageType[]) => void,
+  children: (
+    schedules: ScheduleType[],
+    onAssign: (scheduleId: number, cancelActions: boolean) => Promise<any>
+  ) => Node
+};
+
+export function WithMaintenanceSchedules(props: WithMaintenanceSchedulesProps) {
```
Comment (parlt91) [inline web/html/src/manager/maintenance/ssm/schedule-picker.js:33]: Same thought here maybe move the code in `maintenance/ssm` to `systems/ssm`? On the other hand I am not against having everything maintenance window related in `maintenance/`. I'd like to hear some of your opinions about this.

File: web/html/src/components/dialog/DangerDialog.js [hunk 3]
```diff
@@ -72,4 +73,5 @@ DangerDialog.propTypes = {
     onConfirmAsync: PropTypes.func,
     submitText: PropTypes.string,
     submitIcon: PropTypes.string,
+    btnClass: PropTypes.string
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/components/dialog/DangerDialog.js:76]: All these declarations are optional. With PropTypes, required ones are declared like:
```
btnClass: PropTypes.string.isRequired
```
I know, weird :stuck_out_tongue_winking_eye: 
These older components need a Flow refactoring still.

File: web/html/src/components/buttons.js [hunk 4]
```diff
@@ -65,7 +65,7 @@ export class AsyncButton extends _ButtonBase {
     let style = "btn ";
     switch (this.state.value) {
         case "failure": style += "btn-danger"; break;
-        case "waiting": style += "btn-default"; break;
+        case "waiting":
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/components/buttons.js:68]: If any, it'd fix more cases.
Initially, this was used only with "default" style buttons so we didn't hit the bug.

**Comment by cbbayburt** [comment]
@parlt91 @hustodemon thanks for the reviews!

> There is just one usability thing: in the schedule details -> assign systems, when I assign a system to a schedule, it disappears from the list. Wouldn't it make more sense, to use this UI also for unassigning (or retracting) systems from schedules? Or alternatively, there could be another tabe unassign/retract systems for this purpose (maybe something to be discussed with @ncounter too). This can be addressed in a different PR, if it's considered to be a good idea.

I'll definitely revisit that table after [server-side pagination](https://github.com/SUSE/spacewalk/issues/11895). It can be either a second table with already assigned systems, or a single table with assigned systems are preselected.

File: web/html/src/components/buttons.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hustodemon) [reply web/html/src/components/buttons.js:68]: ah, good! :smile:

File: web/html/src/components/dialog/DangerDialog.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (hustodemon) [reply web/html/src/components/dialog/DangerDialog.js:76]: I see. I should study react a bit. Thanks for explaining.

**Comment by hustodemon** [comment]
> I'll definitely revisit that table after server-side pagination. It can be either a second table with already assigned systems, or a single table with assigned systems preselected.

Great, let's discuss this later!

File: java/code/webapp/WEB-INF/pages/ssm/ssmindex.jsp [hunk 5]
```diff
@@ -173,6 +173,7 @@
                         <rhn:require acl="all_systems_in_set_have_feature(ftr_package_refresh)">
                             <li><bean:message key="ssm.overview.misc.packageprofiles"/></li>
                         </rhn:require>
+                        <li><bean:message key="ssm.overview.schedule.assign"/></li>
```
Comment (ncounter) [inline java/code/webapp/WEB-INF/pages/ssm/ssmindex.jsp:176]: Just  to make sure: no restriction for all the type of user right?

File: web/html/src/core/spa/spa-renderer.js [hunk 6]
```diff
@@ -26,7 +26,10 @@ function hasReactApp() {
   return window.pageRenderers.spa.reactAppsName.length > 0;
 }
 
-function renderGlobalReact(element: ReactElement<any>, container: Element) {
+function renderGlobalReact(element: ReactElement<any>, container: ?Element) {
+  if (container == null) {
```
Comment (ncounter) [inline web/html/src/core/spa/spa-renderer.js:30]: Woah, sounds new to me. What's the scenario did you experienced it?

File: web/html/src/manager/maintenance/ssm/schedule-picker.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  const onAssign = (scheduleId: number, cancelActions: boolean): Promise<any> => {
+    let uri;
+    let data;
+    let successMsg;
+    if (scheduleId === 0) {
+      uri = "/rhn/manager/api/maintenance/schedule/unassign";
+      successMsg = t("Maintenance schedule has been cleared");
+      data = props.systems;
+    } else {
```
Comment (ncounter) [inline web/html/src/manager/maintenance/ssm/schedule-picker.js:44]: Checkstyle: `else` should go to a new line

File: web/html/src/manager/maintenance/ssm/system-assignment.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import React, {useState} from "react";
+import SpaRenderer from "core/spa/spa-renderer";
+
+import {Messages} from "components/messages";
+import {BootstrapPanel} from "components/panels/BootstrapPanel";
+
+import {SchedulePickerForm, WithMaintenanceSchedules} from "./schedule-picker";
+
+function SystemAssignment(props: {systems: string[]}) {
+  // TODO: Add proper info text
```
Comment (ncounter) [inline web/html/src/manager/maintenance/ssm/system-assignment.js:12]: Is the `TODO` still something missing? I'd either address it or drop it.

File: java/code/src/com/redhat/rhn/common/db/datasource/xml/System_queries.xml [hunk 9]
```diff
@@ -2419,6 +2419,17 @@ ORDER BY UPPER(COALESCE(X.SERVER_NAME, '(none)')), X.ID
   <elaborator name="entitlements" />
 </mode>
 
+<mode name="target_systems_for_maintenance_schedule" class="com.redhat.rhn.frontend.dto.EssentialServerDto">
+    <query params="user_id, schedule_id">
+        SELECT s.id, s.name
+        FROM rhnServer s JOIN rhnUserServerPerms usp on s.id = usp.server_id
+        WHERE usp.user_id = :user_id
+            AND (s.maintenance_schedule_id != :schedule_id
```
Comment (ncounter) [inline java/code/src/com/redhat/rhn/common/db/datasource/xml/System_queries.xml:2427]: Maybe better to check first if it is `null` and then by comparing by `schedule_id` just for the sake of performance?

File: java/code/webapp/WEB-INF/pages/ssm/ssmindex.jsp (hunk 5 above)
Comment (cbbayburt) [reply java/code/webapp/WEB-INF/pages/ssm/ssmindex.jsp:176]: IIUC, yes. But as the authors of the RFC, maybe @mcalmer or @hustodemon have some other opinion?

File: web/html/src/core/spa/spa-renderer.js (hunk 6 above)
Comment (cbbayburt) [reply web/html/src/core/spa/spa-renderer.js:30]: We don't hit that case in real-life, but theoretically, this is usually called with `document.getElementById()`. `getElementById` can return `null` if the `id` does not exist and Flow knows about this. So basically, this is to satisfy Flow. I made the check in a single place here so we won't have to check wherever we call `getElementById`.

File: web/html/src/manager/maintenance/ssm/system-assignment.js (hunk 8 above)
Comment (cbbayburt) [reply web/html/src/manager/maintenance/ssm/system-assignment.js:12]: Yes, I need some fancy text to put in here :) I'll take care of it before merging. Suggestions welcome (check out the panel header in the *Assignment from SSM* screenshot)

File: web/html/src/core/spa/spa-renderer.js (hunk 6 above)
Comment (ncounter) [reply web/html/src/core/spa/spa-renderer.js:30]: Awesome, thanks for the explanation.

File: java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java (hunk 1 above)
Comment (cbbayburt) [reply java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java:129]: SSM is generally structured as a collection of operations from various domains and every domain adds it's own functionality from it's own namespace. For example: `ssm/highstate` is served from `MinionController`.

**Comment by ncounter** [comment]
> My comments have been addressed. LGTM, thanks @cbbayburt +1

Same for me. Thanks a lot @cbbayburt

**Pull request merged**